### ISSUE TITLE
Clippy fix replaces `HashMap::or_insert` with `HashMap::or_default` inside a `fold`, breaking type inference

### ISSUE THREAD

**Issue by MaeIsBad** [rust-lang/rust-clippy#12636]
### Summary

Running `cargo clippy --fix` results in code that doesn't compile

### Reproducer

```rust
use std::collections::HashMap;

fn main() {
    let _configs = ['a','b','c']
        .into_iter()
        .fold(HashMap::new(), |mut acc, k| {
            acc.entry(k)
                .or_insert(Vec::new())
                .push(k);
            acc
        });
}
```

Running
`cargo clippy --fix`

results in
```
    Checking asdf v0.1.0 (/home/mae/asdf)
warning: failed to automatically apply fixes suggested by rustc to crate `asdf`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0282]: type annotations needed for `std::collections::HashMap<char, V>`
 --> src/main.rs:6:32
  |
6 |         .fold(HashMap::new(), |mut acc, k| {
  |                                ^^^^^^^
...
9 |                 .push(k);
  |                  ---- type must be known at this point
  |
help: consider giving this closure parameter an explicit type, where the type for type parameter `V` is specified
  |
6 |         .fold(HashMap::new(), |mut acc: std::collections::HashMap<char, V>, k| {
  |                                       ++++++++++++++++++++++++++++++++++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0282`.
Original diagnostics will follow.

warning: use of `or_insert` to construct default value
 --> src/main.rs:8:18
  |
8 |                 .or_insert(Vec::new())
  |                  ^^^^^^^^^^^^^^^^^^^^^ help: try: `or_default()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unwrap_or_default
  = note: `#[warn(clippy::unwrap_or_default)]` on by default

warning: `asdf` (bin "asdf" test) generated 1 warning (run `cargo clippy --fix --bin "asdf" --tests` to apply 1 suggestion)
warning: failed to automatically apply fixes suggested by rustc to crate `asdf`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0282]: type annotations needed for `std::collections::HashMap<char, V>`
 --> src/main.rs:6:32
  |
6 |         .fold(HashMap::new(), |mut acc, k| {
  |                                ^^^^^^^
...
9 |                 .push(k);
  |                  ---- type must be known at this point
  |
help: consider giving this closure parameter an explicit type, where the type for type parameter `V` is specified
  |
6 |         .fold(HashMap::new(), |mut acc: std::collections::HashMap<char, V>, k| {
  |                                       ++++++++++++++++++++++++++++++++++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0282`.
Original diagnostics will follow.

warning: `asdf` (bin "asdf") generated 1 warning (1 duplicate)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s
```

### Version

```text
rustc 1.79.0-nightly (385fa9d84 2024-04-04)
binary: rustc
commit-hash: 385fa9d845dd326c6bbfd58c22244215e431948a
commit-date: 2024-04-04
host: x86_64-unknown-linux-gnu
release: 1.79.0-nightly
LLVM version: 18.1.2
```
